File: python/src/coding_interview/domain/user_id.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class UserId:
    """ユーザーIDを表す。"""

    value: str

    def __post_init__(self) -> None:
        if not self.value:
            raise ValueError("userId must not be empty")

File: ruby/lib/coding_interview/domain/user_id.rb
module CodingInterview
  module Domain
    # ユーザーIDを表す。
    class UserId
      attr_reader :value

      def initialize(value)
        raise ArgumentError, "userId must not be empty" if value.nil? || value.empty?
        @value = value
      end

      def ==(other)
        other.is_a?(UserId) && other.value == @value
      end
      alias eql? ==

      def hash
        @value.hash
      end

      def to_s
        @value
      end
    end
  end
end
